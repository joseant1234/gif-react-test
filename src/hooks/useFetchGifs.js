import { useState, useEffect } from "react"
import { getGifs }  from '../helpers/getGifs';

// los custom hooks funcionan como functional components, puedes tener efectos, reducers, contextos
export const useFetchGifs = (category) => {

    const [state, setState] = useState({
        data: [],
        loading: true,
    });

    // los efectos no pueden ser async (useEffect(async() ), porque esperan ser sincronos
    useEffect(() => {
        getGifs(category).then(imgs => {

            setState({
                data: imgs,
                loading: false,
            });
        })
    }, [category])

    return state;
}
