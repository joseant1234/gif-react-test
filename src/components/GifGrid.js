import React, { useState, useEffect } from 'react'
import { GifGridItem } from './GifGridItem';
// import { getGifs } from '../helpers/getGifs';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {

    // const [images, setImages] = useState([]) ;

    const { data:images, loading } = useFetchGifs(category);



    // con el arreglo vació solo se ejecuta una vez lo que está dentro del effect. Se ejecuta cuando se renderiza por primera vez el componente
    // useEffect(() => {
    //     getGifs(category).then(setImages(imgs));
    // }, [])

    // si se envía en el arreglo de dependecias, cada vez que se actualice category se va a ejecutar el effect
    // useEffect(() => {
    //     getGifs(category).then(setImages);
    // }, [category])

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{ category }</h3>

            { loading && <p className="animate__animated animate__flash">Loading</p> }

            <div className="card-grid">

                {
                    images.map((img) => (
                        <GifGridItem
                            key={img.id}
                            {...img}
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}
