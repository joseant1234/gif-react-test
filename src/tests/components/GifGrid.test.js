import React from 'react';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';

jest.mock('../../hooks/useFetchGifs');

describe('Test in <GifGrid />', () => {

    const category = 'Naruto';

    test('should show correctly', () => {
        // cuando se llame useFetchGifs dentro del componente va enviar el valor puesto como parametro del metodo mockReturn()
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true,
        });
        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();

    });

    test('should show items when images are loading using useFetchGifs', () => {
        const gifs = [{
            id: 'Example',
            url: 'https://localhost/something.jpg',
            title: 'Title',
        }]
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false,
        });

        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();


    })
})
