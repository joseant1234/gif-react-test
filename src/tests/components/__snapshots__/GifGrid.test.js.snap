// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <GifGrid /> should show correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Naruto
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Test in <GifGrid /> should show items when images are loading using useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Naruto
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="Example"
      key="Example"
      title="Title"
      url="https://localhost/something.jpg"
    />
  </div>
</Fragment>
`;
